### PR TITLE
Add support for target paths to build.d

### PR DESCRIPTION
**Pull request by marler8997**
Split up https://github.com/dlang/dmd/pull/9979

Causes `build.d` to fallback to matching command-line arguments to target paths when it doesn't match any name.  Allows user to build specific files, much like Make allows you to do:
```
./build.d generated/linux/release/64/dmd.conf
```

The command line argument only needs to match the end of the path, so the following would also work:
```
./build.d dmd.conf
# also
./build.d 64/dmd.conf
```

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

File: src/build.d [hunk 1]
```diff
@@ -347,6 +353,7 @@ auto predefinedTargets(string[] targets)
 {
     import std.functional : toDelegate;
     Appender!(void delegate()[]) newTargets;
+LtargetsLoop:
```
Comment (JinShil) [inline src/build.d]: Yuck!  Is there some other way?

File: src/build.d [hunk 2]
```diff
@@ -414,6 +433,33 @@ auto predefinedTargets(string[] targets)
     return newTargets.data;
 }
 
+// An input range for a recursive set of dependencies
```
Comment (JinShil) [inline src/build.d]: Please use a DDoc comment.  It makes the comment appear when hovering over a symbol in some developers' editors.

File: src/build.d [hunk 3]
```diff
@@ -414,6 +433,33 @@ auto predefinedTargets(string[] targets)
...
+    this(DependencyRef[] deps) { addDeps(deps); }
+    bool empty() const { return next.length == 0; }
+    auto front() inout { return next[0]; }
+    void popFront()
+    {
+        auto save = next[0];
+        next = next[1 .. $];
+        addDeps(save.deps);
+    }
+    void addDeps(DependencyRef[] deps)
```
Comment (JinShil) [inline src/build.d]: DDoc comment for the benefit of those using editors that support DDoc features.

Comment (JinShil) [reply src/build.d]: Actually, I don't see any need for this to be a function, can't you just put it into the constructor?

File: src/build.d [hunk 4]
```diff
@@ -29,6 +29,11 @@ __gshared string[string] env;
 __gshared string[][string] flags;
 __gshared typeof(sourceFiles()) sources;
 
+immutable rootDeps = [
```
Comment (JinShil) [inline src/build.d]: Please add some DDoc to these to help the next person who has to maintain this code.

File: src/build.d (hunk 1 above)
Comment (marler8997) [reply src/build.d]: Yeah I could a `bool foundMatch = false` ... `foundmatch=true; break` ... `if (!foundMatch) { throw error`....

but it seems alot more complex than just saying, `break LtargetsLoop`.  Note that labeling loops is alot less complex than general labels and gotos.

File: src/build.d (hunk 3 above)
Comment (marler8997) [reply src/build.d]: It's called from the constructor and from popFront

File: src/build.d (hunk 4 above)
Comment (marler8997) [reply src/build.d]: done

File: src/build.d (hunk 2 above)
Comment (marler8997) [reply src/build.d]: done

**Review by JinShil — APPROVED** [review]

**Pull request merged**